src/Type/Spryker/ClassDynamicTypeExtension.php

```diff
@@ -0,0 +1,70 @@
+<?php declare(strict_types = 1);
+
+/**
+ * MIT License
+ * Use of this software requires acceptance of the Evaluation License Agreement. See LICENSE file.
+ */
+
+namespace PHPStan\Type\Spryker;
+
+use PhpParser\Node\Expr\MethodCall;
+use PhpParser\Node\Identifier;
+use PHPStan\Analyser\Scope;
+use PHPStan\Reflection\MethodReflection;
+use PHPStan\Type\DynamicMethodReturnTypeExtension;
+use PHPStan\Type\ObjectType;
+use PHPStan\Type\Type;
+
+abstract class ClassDynamicTypeExtension implements DynamicMethodReturnTypeExtension
+{
+
+	/** @var bool[] */
+	protected $methodResolves = [];
+
+	public function isMethodSupported(MethodReflection $methodReflection): bool
+	{
+		if (isset($this->methodResolves[$methodReflection->getName()])) {
+			return true;
+		}
+
+		return false;
+	}
+
+	/**
+	 * @param \PHPStan\Reflection\MethodReflection $methodReflection
+	 * @param \PhpParser\Node\Expr\MethodCall $methodCall
+	 * @param \PHPStan\Analyser\Scope $scope
+	 *
+	 * @throws \Exception
+	 * @throws \PHPStan\ShouldNotHappenException
+	 *
+	 * @return \PHPStan\Type\Type
+	 */
+	public function getTypeFromMethodCall(MethodReflection $methodReflection, MethodCall $methodCall, Scope $scope): Type
+	{
+		if (!$scope->isInClass()) {
+			throw new \PHPStan\ShouldNotHappenException();
+		}
+
+		if (!$methodCall->name instanceof Identifier) {
+			throw new \PHPStan\ShouldNotHappenException();
+		}
+
+		$docComment = $scope->getClassReflection()->getNativeReflection()->getDocComment();
+
+		if ($docComment === false) {
+			throw new \Exception('Please add PHPDoc block with @method annotation for "getFactory(), getQueryContainer() and/or getFacade()" if one  is used.');
```

Comment:
Extensions should never throw exceptions, unless it's a completely inexpected situations. If you're tempted to throw exceptions from an extension, you should write a corresponding rule instead (and return `MixedType` or `ErrorType` here).